src/state.rs

```diff
@@ -118,19 +145,46 @@ impl VMState {
             };
 
         self.pubdata_costs.entries.push(0);
+        self.refunds.entries.push(refund);
 
         (value, refund)
     }
 
-    fn storage_read_inner(&self, key: &StorageKey) -> Option<U256> {
+    pub fn storage_read_with_no_refund(&mut self, key: StorageKey) -> U256 {
+        let value = self.storage_read_inner(&key).unwrap_or_default();
+        let storage = self.storage.borrow();
+
+        if !storage.is_free_storage_slot(&key) && !self.read_storage_slots.map.contains(&key) {
+            self.read_storage_slots.map.insert(key);
+        };
+
+        self.pubdata_costs.entries.push(0);
+
+        value
+    }
+
+    fn storage_read_inner(&mut self, key: &StorageKey) -> Option<U256> {
         match self.storage_changes.map.get(key) {
-            None => self.storage.borrow_mut().storage_read(key),
+            None => {
+                let cached_value = *self.initial_values.get(key).unwrap_or(&None);
+                if cached_value.is_none() {
+                    let value = self.storage.borrow_mut().storage_read(key);
```

Comment:
Yes, because we are adding a caching layer for the storage, to not query the db in the `zksync-era`, since we read a lot, specially when getting diffs.